include/envoy/common/account.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+
+class Charge;
+
+/*
+ * An interface for accounting the usage of a resource.
+ * The "resource" tracked here is implicit, if in the future we want to track multiple
+ * resources, we should expand the interface to make it explicit.
+ *
+ * Currently this is only used by L7 streams to track the amount of memory used
+ * by the stream.
+ *
+ */
+class Account {
+public:
+  virtual ~Account() = default;
+
+  /**
+   * Returns the outstanding balance.
+   */
+  virtual uint64_t balance() const PURE;
+
+  /**
+   * Charges the account the specified amount.
+   *
+   * @param amount the amount to debit.
+   * @return the charge object that when destroyed, will credit the account
+   * freeing the resource.
+   */
+  virtual std::unique_ptr<Charge> charge(uint64_t amount) PURE;
```

Comment:
That's a good point. I think, given that the actual account object is fairly small, keeping it around with a `shared_ptr` should be alright. Will switch to using `shared_ptr` generally instead, then I can add those asserts. 

Furthermore, I'll make the credit method public. I think we can push charge functionality more into the `Slice` itself, removing the `Charge` class.

Slices will just track via `shared_ptr` the accounts themselves which it can credit on dtor.